MobileSdk/src/main/java/com/spruceid/mobile/sdk/GattClient.kt

```diff
@@ -333,101 +348,300 @@ class GattClient(private var callback: GattClientCallback,
-            usingL2CAP = characteristicL2CAP != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q
-
-            if (usingL2CAP) {
-                callback.onLog("Using L2CAP: $usingL2CAP")
+    private fun readRequest() {
+        val payload = ByteArrayOutputStream()
+
+        // Keep listening to the InputStream until an exception occurs.
+        val inStream = try {
+            l2capSocket!!.inputStream
+        } catch (e: IOException) {
+            reportError("Error on listening input stream from socket L2CAP: ${e}")
+            return
+        }
 
-                // value is returned async above in onCharacteristicRead()
-                if (!gatt.readCharacteristic(characteristicL2CAP)) {
-                    callback.onError(Error("Error reading L2CAP characteristic."))
+        while (true) {
+            val buf = ByteArray(L2CAP_BUFFER_SIZE)
+            try {
+                val numBytesRead = inStream.read(buf)
+                if (numBytesRead == -1) {
+                    reportError("Failure reading request, peer disconnected.")
+                    return
                 }
+                payload.write(buf, 0, buf.count())
+
+                dprint("Currently have ${buf.count()} bytes.")
+
+                requestTimestamp = TimeSource.Monotonic.markNow()
+
+                Executors.newSingleThreadScheduledExecutor()
+                    .schedule({
+                                  val curtime = TimeSource.Monotonic.markNow()
+                                  if ((curtime - requestTimestamp) > 500.milliseconds) {
+                                      val message = payload.toByteArray()
+
+                                      dprint("Request complete: ${message.count()} bytes.")
+                                      callback.onMessageReceived(message)
+                                  }
+                              }, 500, TimeUnit.MILLISECONDS)
+
+            } catch (e: IOException) {
+                reportError("Error on listening input stream from socket L2CAP: ${e}")
                 return
             }
+        }
+    }
+
+    /**
+     * Thread for writing the response via L2CAP.
+     */
+    fun writeResponse() {
+        val outStream = l2capSocket!!.outputStream
+        try {
+            while (true) {
+                var message: ByteArray?
+                try {
+                    message = responseData.poll(500, TimeUnit.MILLISECONDS)
+                    if (message == null) {
+                        continue
+                    }
+                    if (message.size == 0) {
+                        break
+                    }
+                } catch (e: InterruptedException) {
+                    continue
+                }
+                outStream.write(message)
+            }
+        } catch (e: IOException) {
+            reportError("Error writing response via L2CAP socket: ${e}")
+        }
+
+        try {
+            // TODO: This is to work around a bug in L2CAP
```

Comment:
can you also leave a link or a short description on what the bug is? just for further reference if someone else gets to fix it.

Reply:
Done.